### PR TITLE
Farming Tracker Plugin

### LINKED ISSUE #511: Farming Timer Widget Similar to OSB/Konduit
This is the only thing stopping me from using runelite. It keeps track of herbs/trees etc on the right side using a Widget. Possible addition of a notification when your plants are ready would be nice.

### PR DESCRIPTION
**Pull request by Abextm**
Adds a farming tracker that uses varbits to store state. This supports
 - Multiple accounts
 - Tracking planted produce
 - Growth time estimation

I'm reasonably sure there are still bugs in some of the patch implementations.

Depends on #1404 
Supersedes #1244
Fixes #511

### CODE REVIEW

**Comment by deathbeam** [comment]
Just small UI related comment, but that layout in the screenshot looks uhh, horrible :d

File: runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPlugin.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	@Override
+	protected void startUp() throws Exception
+	{
+		BufferedImage icon;
+		synchronized (ImageIO.class)
+		{
+			icon = ImageIO.read(getClass().getResourceAsStream("farming.png"));
+		}
+
+		panel = injector.getInstance(FarmingTrackerPanel.class);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPlugin.java]: Don't create swing components through Guice.

File: runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPanel.java [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+	private FarmingWorld farmingWorld;
+
+	private List<FarmingPatchPanel> patchPanels = new ArrayList<>();
+
+	FarmingTrackerPanel()
+	{
+		super(false);
+	}
+
+	void init()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPanel.java]: Merge this method with constructor and dont create Swing compoents through Guice

File: runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPlugin.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+		{
+			return;
+		}
+
+		FarmingRegion region = farmingWorld.getRegions().get(loc.getRegionID());
+		if (region != null)
+		{
+			// Write config with new varbits
+			// farmingTracker.<login-username>.<regionID>.<VarbitID>=<varbitValue>:<unix time>
+			String group = FarmingTrackerConfig.KEY_NAME + "." + client.getUsername() + "." + region.getRegionID();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPlugin.java]: This seems like workaround until proper user-based settings are added? I don't think this should be done and just leave it be just like SlayerPlugin works atm

File: runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingPatchPanel.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import java.awt.Dimension;
+import javax.swing.GroupLayout;
+import javax.swing.JLabel;
+import javax.swing.JPanel;
+import javax.swing.JProgressBar;
+import lombok.Getter;
+
+class FarmingPatchPanel extends JPanel
+{
+	@Getter
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingPatchPanel.java]: You can just annotate class with `@Getter` here

File: runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerConfig.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import net.runelite.client.config.ConfigItem;
+
+@ConfigGroup(
+	keyName = "farmingTracker",
+	name = "Farming Tracker",
+	description = "Configuration for the farming tracker"
+)
+public interface FarmingTrackerConfig extends Config
+{
+	String KEY_NAME = "farmingTracker";
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerConfig.java]: Can't you use this KeyName in annotation?

File: runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPanel.java [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+					if (doneEstimate < unixNow)
+					{
+						panel.getEstimate().setText("Done");
+					}
+					else if (config.estimateRelative())
+					{
+						int remaining = (int) (59 + doneEstimate - unixNow) / 60;
+						StringBuilder f = new StringBuilder();
+						f.append("Done in ");
+						int min = remaining % 60;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPanel.java]: There must be some date fromatting function ready to use, no neet to reinvent the wheel here

**Comment by Kamielvf** [comment]
When I cleared one of my patches it would show the following, which I assume would be correct but I have the auto-weed active so the weed will never grow.
This is me changing auto-weed from disabled to enabled:
`Widget setting index 1329 has changed from 4194354 to 8388658: 00000000 01000000 00000000 00110010 -> 00000000 10000000 00000000 00110010`
image

When moving between the falador crops/herb patch region and the port sarim spirit tree, the following happens:
image

The lumbridge tree patch has the wrong region bound to it, and after changing it to the correct one it still doesn't detect it properly:
image

File: runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPlugin.java (hunk 3 above)
Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPlugin.java]: I've had multiple people ask me if this was going to work for multiple accounts, and I don't see proper per-user settings coming for at least a couple weeks, so I would like to keep this behavior.

File: runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerConfig.java (hunk 5 above)
Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerConfig.java]: No

File: runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPanel.java (hunk 6 above)
Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPanel.java]: I couldn't find a implementation in the standard library that would elide zero-valued segments

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/farmingtracker/FarmingTrackerPanel.java]: I see

**Comment by Abextm** [comment]
UI has been updated to look a little nicer.

**Pull request merged**